RFC for account deletion

rbonichon opened the pull request:
This document describes the account deletion feature, its motivation and proposed solution as well as possible culprits.

rbonichon committed:
account_deletion: skeleton

rbonichon committed:
account deletion: extract some contents from PRD

rbonichon committed:
account deletion: design considerations

rbonichon committed:
Some details on ledger implementation

rbonichon committed:
Account deletion: some elements of testing

rbonichon committed:
Some further notes on ledger API

rbonichon committed:
Precisions

rbonichon committed:
Choice

rbonichon committed:
some tentative words about transaction logic

this is not clear yet

mrmr1993 commented on 00xx-account-deletion.md:
````diff
@@ -0,0 +1,272 @@
+new insertions
+
+```
+    H(H(A,B),H(C,D))       free = [[Right; Left]]
````
Indexes are already dense, shouldn't we use those?

rbonichon replied:
Do you mean using the `Bigstring` view of the same information, that is, `Merkle_address` ?

rbonichon replied:
The actual implementation will use `Location.t`, which may or may not be  a compact representation since we are within the functor's body. The current instance of this functor does use a compact representation for indexes. I chose the `Right, Left` way of wording the location because I thought it was easier to follow, especially in a tree-like drawing.

mrmr1993 reviewed (changes requested):
I don't think this digs into the mechanisms in the way I was expecting. Can you add some detail about what needs to be implemented to achieve the described behaviour.

rbonichon committed:
account deletion: update

- more precision in Merkle tree explanation
- expand upond transaction logic

rbonichon committed:
account deletion: update

- more precision in Merkle tree explanation

mrmr1993 reviewed (changes requested):
Can you update this to be in line with [the template](https://github.com/o1-labs/rfcs/blob/main/0001-template.md)? More comments to follow

mrmr1993 commented on 00xx-account-deletion.md:
````diff
@@ -0,0 +1,352 @@
+  A    B     C    D
+```
+
+Upon removal of `C`, the structure would evolve as follows, with `x` a
````
This is the only mention of this placeholder value. The details here seem important, e.g. we don't want a deleted account to continue to use the same amount of memory, otherwise we shouldn't refund the creation fee.

rbonichon replied:
Indeed. Let me be more clear here.

rbonichon replied:
Would you maybe prefer also to use symbol `o` in place of `x` since it actually is meant to mark the empty account

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,352 @@
+
+The `Database` module implements an on-disk ledger that backs up the in-memory data structure.
+
+Deletion support relies on the same techniques as in-memory ledgers, tailored to the on-disk db.
```
What does this mean?

rbonichon replied:
Updated

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,352 @@
+updates, i.e., extend the `Account_update.Body.Stable.t` type with an additional
+record field `delete_account`.
+
+This turns the actual definition from
```
Repeating the type twice here feels like redundant padding. Delete.

rbonichon replied:
Done

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,352 @@
+
+This extension provides is thus a version 2 of the current versioned type (`V2`).
+
+What about the changes in Mina_wire_Types?
```
The transactions we send over the wire will change, surely

rbonichon replied:
Sure. I think this question (to myself actually) was more: could that be an issue?

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,352 @@
+deletion.
+
+Clearly, account deletion should not be permitted except for authorized keys, e.g., the
+contract that create the consumable account. For that, we need to check this
```
Check what? Be concrete

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,352 @@
+What about the changes in Mina_wire_Types?
+
+Account deletion is also a payment of the initial account
+creation fee back to an address with extended with that triggers account
```
Garbled sentence

rbonichon replied:
Fixed

rbonichon committed:
Reorganized and add sections to follow template more closely

mrmr1993 commented on 00xx-account-deletion.md:
````diff
@@ -0,0 +1,367 @@
+  A    B     E    D
+```
+
+##### Option 2: Maintain insertion on leftmost available location
````
Alternative, dump it in the alternatives section.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,367 @@
+   <-> account mappings;
+*  account locations in this representation are fixed for the lifetime of the
+  account: proof of inclusion would not change but for the last element of the
+  Merkle path.
```
Alternatives section

rbonichon replied:
Done

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,367 @@
+We propose to support removing elements in ledgers through  2 functions :
+
+- `val remove_location: t -> location -> unit`
+- `val remove_account: t -> account -> unit`
```
What do these do?

rbonichon replied:
Added specification

rbonichon committed:
account deletion : make the meaning `x` more explicit

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,367 @@
+
+The support for deletion at the ledger level now needs to be lifted within the
+transaction logic for account updates.
+The key idea here is to handle removal as a specific kind of account update for
```
There's no mention of the snark-related code here. That will be crucially important to get right.

mrmr1993 replied:
How does the transaction logic differ from what's present today? E.g.
* How should the API to signal that we're deleting an account look?
* How should the native version of that work?
* How should the snark version of that work?
* What happens in the case of failure? Is it already handled?
* How is the global supply of Mina tracked faithfully?

rbonichon committed:
account deletion : preliminary words on account deletion

rbonichon committed:
account deletion : move sections to rationale and alternatives

rbonichon committed:
account deletion : a word on archive nodes

rbonichon committed:
account deletion : drawback impact

rbonichon committed:
account deletion : remove redundant definition

resolves https://github.com/o1-labs/rfcs/pull/51#discussion_r1652807682

rbonichon committed:
account deletion : cleaning whitespaces

rbonichon committed:
account deletion : rewording paragraph

Partly in reaction to
https://github.com/o1-labs/rfcs/pull/51#discussion_r1652809689

rbonichon committed:
account deletion : specify behaviors of `remove_` functions

Initial response to
https://github.com/o1-labs/rfcs/pull/51#discussion_r1652855860

rbonichon committed:
account deletion : typo

this -> these

mrmr1993 commented on 00xx-account-deletion.md:
````diff
@@ -0,0 +1,384 @@
+record field `delete_account`.
+
+This turns the current definition into
+    ```ocaml
````
Indentation here breaks the markdown formatting.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,384 @@
+
+To illustrate option 1, let us assume we have the following Merkle tree of depth
+2, with an empty free list. This Merkle tree has one location marked with `x` -
+this marks the empty account.
```
I don't think this is the design we want in practice: if we just replace it with the empty account, we don't actually save any space. We would really like to delete the account entirely (at least, on disk).

rbonichon replied:
I see your point. We don't really need to store the empty account though, we just need to act as if it were stored in the leaf. I will rephrase that bit.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,384 @@
+
+The view of the free list above is a simplification.
+
+The data contained in the free list needs to be Merkelized as well.  This means,
```
> needs to be Merkelized as well

Please propose a usage if we're going to this effort. E.g. do we need to store it alongside the ledger hash now? Do we need to update the snark ledger update mechanism to account for it? How do we handle the empty merkle list in snark, where the account index is currently non-deterministic? How does this impact ledger sync?

rbonichon replied:
Done

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,384 @@
+
+> Here are some things I am thinking about, maybe they help you inform the RFC:
+>
+> - easy API - we need to express account deletion via an easy API in the sdk
```
The SDK should only be interfacing with the `Account_update.t` portion, so the API is 'handling and extra boolean' at that level, plus some wrapping TBD.

rbonichon replied:
This part should be commented out. This was just for me to check that these points are addressed in the RFC

rbonichon committed:
account deletion : fix indentation

In response to https://github.com/o1-labs/rfcs/pull/51#discussion_r1653040944

rbonichon committed:
account deletion: light clarifications

rbonichon committed:
account deletion: questions and things to consider

rbonichon committed:
account deletion: syncing ledgers

rbonichon committed:
account deletion: word on code links.

rbonichon committed:
account deletion: inital wording on snark

rbonichon committed:
account deletion: implementation steps

rbonichon committed:
account deletion: transaction snark

- missing the specification of the `remove_account` function

rbonichon committed:
account deletion: ledger sync

rbonichon committed:
account deletion: db encoding for free list

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,471 @@
+In turn, the ledger state is the Merklized pair of the ledger itself (as a
+Merkle tree) and its free list.
+
+TBD: using an ordered free list to avoid Merklization
```
Please add these details, and choose between the options, punting the other to the alternatives section.

rbonichon replied:
Done

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,471 @@
+
+<!-- This needs to be rewritten, not sure this is correct -->
+We will implement a decoding function where one can specify how many locations
+we ideally want to retrieve (up to `n`) to handle functions `set_location_batch`.
```
Don't we have to reserialise the whole thing anyway once we've removed these from the list? The data also grows linearly with the number of free'd accounts, can you justify why that's okay and/or propose an alternative?

rbonichon replied:
The idea here would have been to deserialize/serialize only once instead of `n` times when do batch allocations.

rbonichon replied:
Done

mrmr1993 reviewed (commented):
Please add some details on how this affects the merkle masks.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,471 @@
+At the snark level, deletion should thus be equivalent to proving two things
+1. The location $l$ of account $a₀$ is now  equal to the empty account.
+   A simple implementation is to call [`set_account`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/transaction_snark/transaction_snark.ml#L1531) to location $l$ with the empty account;
+2. The change in MINA of $a₁$ should be equal to
```
This is implicit in 1.

rbonichon replied:
I'm not sure I get your point. How does setting the location of an account a to the empty account implicitly mean that another account b should get some amount of MINA?

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,471 @@
+We propose to restrict this capacity to the account to be deleted, so that only
+yourself can call for your own deletion.
+
+For that, we need to check this with respect to the `permissions` field of
```
This sounds like you're adding a new permission to accounts. Please document.

rbonichon replied:
The last version addresses permissions.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,471 @@
+yourself can call for your own deletion.
+
+For that, we need to check this with respect to the `permissions` field of
+`Account_update.Update.t`.
```
This record declares an update to the permissions, not the permissions themselves.

rbonichon replied:
Corrected

rbonichon committed:
account deletion: free list representation

- add implementation choice
- discuss alternative

rbonichon committed:
account deletion: deallocation at fill frontier

rbonichon committed:
account deletion: other suggestions for free list in db

rbonichon committed:
account deletion: permissions

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,576 @@
+[bheap](https://opam.ocaml.org/packages/bheap/) or
+[CCHeap](https://c-cube.github.io/ocaml-containers/last/containers/CCHeap/index.html))
+for this implementation, that is, the free list is an ordered set of locations,
+from which we always allocate the biggest one.
```
Why are we choosing 'biggest' here? My intuition is that, especially if this feature is widely used, we might have accounts created and deleted at the right side of the tree; it would be natural to treat this as moving the fill frontier backwards. For example, if we had a series of deletes like
```
[a, b, c, d, e, f]
[a, b, c, X, e, f]
[a, b, c, X, e, X]
[a, b, c, X, X, X]
```
then it's natural to treat the states as
```
[a, b, c, d, e, f]
[a, b, c, X, e, f]
[a, b, c, X, e]
[a, b, c]
```
and recover the memory. However, in the current description, an insertion would give us
```
[a, b, c, X, X, g]
```
where in a sense we're always trying to 'waste' the maximum possible number of spaces by choosing the largest. If we instead choose the smallest, it really does behave like we've reduced the fill frontier, and always gives us the option to do that as a performance optimisation.

rbonichon replied:
I explain this in the example right below in the document. There are 2 elements for allocations in the structure:
1. The free list (indices between [0, fill frontier index[), let's call it l
2. The fill frontier index, let's call it i

The strategy would be to always use the free list first (from the biggest to the lowest index), then the fill frontier index if the the free list is empty. 

Replaying your example, this would give 
```
[a, b, c, d, e, f]       l = [] i = 5
[a, b, c, X, e, f]      l = [3]  i = 5
[a, b, c, X, e, X]     l = [3] i = 4
[a, b, c, X, X, X]     l = []  i = 2 ( this is where it's important to have a max heap, to make this reduction easy, since you just have to peek and pop)
```
Thus the next insertion would give 
`[ a, b, c, g]`as you described since the free list has been emptied.

rbonichon replied:
The scenario you describe was supposed to be addressed by the paragraph starting with *Deallocation at fill frontier*

mrmr1993 replied:
There are 2 things in conflict here: coalescing the free space to the right while also inserting to the left to keep it dense. You've addressed one, I was addressing the other. Please talk to both in the RFC. It seems as though either option probably require us to deserialise the whole list (unless we have head and bottom peeking).

rbonichon replied:
Creating then deleting an account at the fill frontier is already correctly handled [here](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_mask/masking_merkle_tree.ml#L497) by just moving the `current_location` (what I call the fill frontier). My understanding is this is what you mean by "coalescing the free space to the right".  

I think this step in the example 
```
[a, b, c, X, e, f]      l = [3]  i = 5
[a, b, c, X, e, X]     l = [3] i = 4
```
coalesces the free space to the right (i moves from 5 to 4, the free list, i.e, the left,  is not updated)
This other step 
```
[a, b, c, X, e, X]     l = [3] i = 4
[a, b, c, X, X, X]     l = []  i = 2
```
shows how the free list and the fill frontier interact with one another to keep the free space to the right ¨coalesced¨. 

I'm not sure what I am missing. Could you show me another example of where what I am proposing doesn't offer a good solution?

mrmr1993 replied:
```
[a, b, c, X, X, f, X]     l = [4, 3] i = 4
[a, b, c, X, e, f, X]     l = [3]  i = 2
```
This is bad.

rbonichon replied:
No in your example `i` would still be 4. Allocating from the free list does not change the fill frontier.

rbonichon replied:
In reverse, deallocation from the fill frontier might affect the free list, like in the example above

rbonichon replied:
I will add a section on allocation strategy to make that clearer. It does not seem to be.

rbonichon replied:
The contents of this discussion is reflected in the document

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,576 @@
+Other strategies are discussed [here](#free_list_db alt).
+
+
+#### Syncing ledgers
```
As well as synchronising ledgers, we also need to consider node startup for genesis. Consider the `ledger` field in the runtime configuration: we need a way to represent that an account is empty in that format, and faithfully convert it, so that the genesis ledger after a hard-fork faithfully represents the state of the ledger once we have account deletion.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,576 @@
+the data on the KV database as follows:
+- the free list is identified by the key `free_list` (like
+  [`last_account_location`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml#L235));
+- store the serialized encoding of the heap in the database.
```
Please define this.

rbonichon replied:
Done

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,576 @@
+5. reserialize the new value of the heap.
+
+**Simple improvement on batch allocations**. Furthermore, we will implement a
+decoding function where one can specify how many locations we ideally want to
```
Presumably this isn't useful in practise: usually, you'll be grabbing the free list so that you can update it to consume a location, so you'll need to reserialise the free list once the head elements (or tail elements as currently written here) have been removed.

It's also useful to mention performance: is there any risk that a large number of free'd accounts creates a free list that slows down accesses, because inherently we're writing a larger and larger list into the database as it grows?

rbonichon replied:
We do use batch allocations in the code. To be sure, I don't know whether the code is executed in practice.

mrmr1993 replied:
We do, but the point is that those batch allocations immediately result in a write of the new data. That is to say: deserialising just the head(s) is usually not what you want, given that you also want the tail so that you can reserialise it.

rbonichon replied:
I tried to explain why it might be useful. We do have to serialize some data back to be written (the tail). For batch allocations, I think it's useful to be able to get `n` locations at once to avoid `n-1` read/deserialize/serialize/write` sequences.

mrmr1993 replied:
I agree, but at a certain point it's simpler to deserialise the whole list, take as many locations as are needed, push any new locations onto the list, and then reserialise the final list.

rbonichon replied:
I agree too. That brings the following question: Is there a scenario where we would in one call allocate mutliple and free multiple locations at the same time?

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,576 @@
+### Updating the archive node
+
+The addition of a field in `Account_update` will need to reflected in table
+`zkapp_account_update_body` with the addition of a `delete_account` column.
```
Add details for the permission.

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,576 @@
+- `val remove_location: t -> location -> unit`: remove the account found at the
+  given `location` from the ledger, don't do anything if `location` is occupied
+  by the empty account
+- `val remove_account: t -> account -> unit`: remove `account` from the ledger,
```
What is `account` here? `account_id` perhaps? Or do you pass a full account that might be in the ledger?

rbonichon replied:
It's an abstract `account` type as defined [here](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/intf.ml#L278). 

Come to think of it an `account_id`might be better.

rbonichon replied:
Updated

rbonichon committed:
account deletion: update remove function description

- in response to https://github.com/o1-labs/rfcs/pull/51#discussion_r1661297673

rbonichon committed:
account deletion: fix fill frontier index definition

rbonichon committed:
account deletion: deallocation at fill frontier revisited

- use response to https://github.com/o1-labs/rfcs/pull/51/files/688cac5958cdf4b6452d784dba03698905a714ae#r1661283831

rbonichon committed:
account deletion: serialization of heap

rbonichon committed:
account deletion: additional question

rbonichon committed:
account deletion: on-disk free list handling

A more thorough discussion about where storing the free list on-disk might fail.

rbonichon committed:
account deletion: JSON encoding of ledger

rbonichon committed:
fixup! account deletion: JSON encoding of ledger

rbonichon committed:
account deletion: some implementation details for permissions

rbonichon committed:
Punctuation

rbonichon committed:
account deletion: remove empty section

rbonichon committed:
account deletion: add detail for permissions in archive nodes

rbonichon committed:
account deletion: some concepts

rbonichon committed:
account deletion: a word about remove account

rbonichon committed:
account deletion: use fill frontier/interval consistently

rbonichon committed:
account deletion: example implementation of remove_account

rbonichon committed:
account deletion: modifying deserialization

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,712 @@
+
+Deleting an account $a₀$ results in two actions:
+1. Actual removal of account $a₀$ of the storage layer;
+2. Return of balance of MINA tokens from $a₉$ and initial creation fee to a specified $a₁$ account.
```
This isn't how we want it in practice. Instead, we want to return the balance to the 'floating balance' for the zkApp txn, and the rest of the txn logic handles the other account. We *should not* specify the `a_1` account (and indeed that might be several accounts in actual usage).

nholland94 replied:
What @mrmr1993 says here is correct. Luckily, it's very easy (in fact, easier than what is described in the RFC as of now): you just need to track the initial creation fee in the fee excess, no other changes will be required.

rbonichon replied:
Updated the document to reflect this discussion

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,712 @@
+for some authorized participants, e.g., the contract that created the consumable
+account. We will extend the [permissions
+type](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/mina_base/permissions.ml#L343)
+with a dedicated `delete` field.
```
Default value?

rbonichon replied:
`Signature`. Done

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,712 @@
+
+Within the protocol-related codebase, the lowest-level support lies in having a
+working and sound `remove` functions for the different ledger implementations
+(`Database`, `Any_ledger`, `Null_ledger`, `Syncable_ledger`) and in masking
```
I think we need more info here about `masking merkle trees`. In particular, it's important to specify the `commit` method to ensure that we're atomically updating all of the relevant information when we e.g. push to disk or apply a zkApp transaction.

rbonichon replied:
Addressed

rbonichon committed:
account deletion: default value for delete permissions

rbonichon committed:
account deletion: gain in batch allocation

Explictly state what we gain when permitting batch allocations.

rbonichon committed:
account deletion: parent/child registering and committing

rbonichon committed:
account deletion: rephrase

mrmr1993 commented on 00xx-account-deletion.md:
```diff
@@ -0,0 +1,732 @@
+commit
+function](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_mask/masking_merkle_tree.ml#L617)
+will need to push its removed locations to its parent atomically, that is, after
+the `is_committing` flage has been set to true
```
`s/flage/flag`

rbonichon replied:
Fixed

rbonichon committed:
account deletion: update discussion on data representation

rbonichon committed:
account deletion: fix discussion

rbonichon committed:
account deletion: update on-disk serializing with sets

rbonichon committed:
account deletion: transaction logic update

Update the text to try to answer comments in
https://github.com/o1-labs/rfcs/pull/51#discussion_r1662875699

rbonichon committed:
account deletion: implementation steps made clearer
